// @progress/kendo-react-conversational-ui/ai-prompt/AIPromptHeader.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import o from "react";
import { Toolbar as c, Button as m } from "@progress/kendo-react-buttons";
const u = (l) => {
  const { activeView: a, toolbarItems: r, activeViewChange: t } = l, n = (e) => {
    t == null || t.call(null, e);
  };
  return /* @__PURE__ */ o.createElement("div", { className: "k-prompt-header" }, /* @__PURE__ */ o.createElement(c, { keyboardNavigation: !1, role: "toolbar", className: "k-toolbar-flat" }, r.map(
    (e) => o.isValidElement(e) ? e : /* @__PURE__ */ o.createElement(
      m,
      {
        key: e.name,
        className: a === e.name ? "k-selected" : "",
        size: "medium",
        themeColor: "primary",
        fillMode: "flat",
        rounded: "full",
        title: e.buttonText,
        svgIcon: e.buttonIcon,
        onClick: () => n(e.name)
      },
      e.buttonText
    )
  )));
};
export {
  u as PromptHeader
};
