// @progress/kendo-react-grid/columnMenu/adaptiveContent/GridActionSheetFooter.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import t from "react";
import { Button as n } from "@progress/kendo-react-buttons";
const a = (e) => /* @__PURE__ */ t.createElement("form", { onSubmit: e.onApply, onReset: e.onReset }, /* @__PURE__ */ t.createElement("div", { className: "k-actions k-actions-stretched" }, /* @__PURE__ */ t.createElement(n, { svgIcon: e.resetIcon, size: "large", type: "reset" }, e.resetMessage), /* @__PURE__ */ t.createElement(
  n,
  {
    disabled: e.submitDisabled,
    svgIcon: e.submitIcon,
    type: "submit",
    size: "large",
    themeColor: "primary"
  },
  e.submitMessage
)));
export {
  a as GridActionSheetFooterContent
};
