// @progress/kendo-react-intl/index.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { GlobalizationContext as e } from "./globalization/GlobalizationContext.mjs";
import { messages as i } from "./Localization/messages.mjs";
import { provideIntlService as l, provideLocalizationService as n, registerForIntl as p, registerForLocalization as s } from "./intlUtils.mjs";
import { ServerGlobalization as x } from "./server/ServerGlobalization.mjs";
import { IntlProvider as c } from "./Intl/IntlProvider.mjs";
import { IntlService as z } from "./Intl/IntlService.mjs";
import { load as g } from "./Intl/load.mjs";
import { loadMessages as L } from "./Localization/loadMessages.mjs";
import { LocalizationProvider as b } from "./Localization/LocalizationProvider.mjs";
import { LocalizationService as F } from "./Localization/LocalizationService.mjs";
import { useInternationalization as M } from "./hooks/useInternationalization.mjs";
import { useLocalization as C } from "./hooks/useLocalization.mjs";
export {
  e as GlobalizationContext,
  c as IntlProvider,
  z as IntlService,
  b as LocalizationProvider,
  F as LocalizationService,
  x as ServerGlobalization,
  g as load,
  L as loadMessages,
  i as localizationMessages,
  l as provideIntlService,
  n as provideLocalizationService,
  p as registerForIntl,
  s as registerForLocalization,
  M as useInternationalization,
  C as useLocalization
};
