// @progress/kendo-react-grid/messages/index.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
const e = "grid.noRecords", t = "grid.pagerInfo", r = "grid.pagerFirstPage", o = "grid.pagerPreviousPage", l = "grid.pagerNextPage", i = "grid.pagerLastPage", a = "grid.pagerItemsPerPage", n = "grid.pagerPage", s = "grid.pagerPageSizeAriaLabel", g = "grid.pagerOf", d = "grid.pagerTotalPages", p = "grid.groupPanelEmpty", u = "grid.groupColumn", c = "grid.ungroupColumn", C = "grid.columnMenu", f = "grid.filterApplyButton", m = "grid.filterClearButton", h = "grid.filterClearAllButton", A = "grid.filterResetButton", b = "grid.filterSubmitButton", I = "grid.filterTitle", P = "grid.sortAscending", S = "grid.sortDescending", T = "grid.sortClearButton", B = "grid.sortApplyButton", O = "grid.searchPlaceholder", y = "grid.searchboxPlaceholder", D = "grid.exportPDF", F = "grid.filterCheckAll", v = "grid.filterChooseOperator", x = "grid.filterSelectAll", E = "grid.filterSelectedItems", L = "grid.sortAriaLabel", R = "grid.editDialogTitle", G = "grid.editDialogSaveButtonTitle", M = "grid.editDialogCancelButtonTitle", w = "grid.filterAriaLabel", q = "grid.groupPanelAriaLabel", N = "grid.groupExpand", k = "grid.groupCollapse", W = "grid.groupClearButton", z = "grid.groupApplyButton", U = "grid.detailExpand", j = "grid.detailCollapse", H = "grid.selectRow", J = "grid.gridAriaLabel", K = "grid.gridRowReorderAriaLabel", Q = "grid.gridAdaptiveColumnMenuFilterTitle", V = "grid.columnMenuColumnChooserTitle", X = "grid.columnMenuColumnChooserSelectedItems", Y = "grid.adaptiveColumnMenuChooserTitle", Z = "grid.adaptiveColumnMenuChooserSubTitle", _ = "grid.columnChooserApplyButton", $ = "grid.columnChooserResetButton", ee = "grid.adaptiveColumnMenuCheckboxFilterTitle", te = "grid.adaptiveToolbarSortTitle", re = "grid.adaptiveToolbarGroupTitle", oe = "grid.toolbarSort", le = "grid.toolbarAI", ie = "grid.aIResponseData", ae = "grid.generatedWithAI", ne = "grid.toolbarAIApply", se = "grid.toolbarGroup", ge = "grid.toolbarFilter", de = "grid.toolbarColumnsChooser", pe = "grid.toolbarCheckboxFilter", ue = {
  [U]: "Expand detail row",
  [j]: "Collapse detail row",
  [N]: "Expand group",
  [k]: "Collapse Group",
  [W]: "Clear grouping",
  [z]: "Done",
  [e]: "No records available",
  [p]: "Drag a column header and drop it here to group by that column",
  [f]: "Apply",
  [m]: "Clear",
  [h]: "Clear all filters",
  [A]: "Reset",
  [b]: "Filter",
  [I]: "Filter",
  [P]: "Sort Ascending",
  [S]: "Sort Descending",
  [T]: "Clear sorting",
  [B]: "Done",
  [L]: "Sortable",
  [te]: "Sort by",
  [re]: "Group by",
  [oe]: "Sort",
  [le]: "AI Assistant",
  [ne]: "Apply",
  [ie]: `Operation is successful. Data is: 
`,
  [ae]: "Generated with AI",
  [se]: "Group",
  [ge]: "Filter",
  [de]: "Columns",
  [pe]: "Filter",
  [u]: "Group Column",
  [c]: "Ungroup Column",
  [C]: "Column menu",
  [a]: "items per page",
  [t]: "{0} - {1} of {2} items",
  [r]: "Go to the first page",
  [o]: "Go to the previous page",
  [l]: "Go to the next page",
  [i]: "Go to the last page",
  [n]: "Page",
  [s]: "Page size",
  [g]: "of",
  [d]: "{0}",
  [O]: "Search",
  [y]: "Search...",
  [D]: "Export PDF",
  [F]: "Check All",
  [R]: "Edit Dialog",
  [G]: "Save",
  [M]: "Cancel",
  [v]: "Choose Operator",
  [E]: "selected items",
  [x]: "Select All",
  [w]: "Filter",
  [q]: "Group panel",
  [J]: "Table",
  [K]: "Drag row",
  [H]: "Select Row",
  [V]: "Columns Chooser",
  [Q]: "Filter by",
  [Y]: "Columns Chooser",
  [Z]: "Selected fields are visible",
  [ee]: "Filter by",
  [X]: "Selected items",
  [_]: "Apply",
  [$]: "Reset",
  "grid.filterEqOperator": "Is equal to",
  "grid.filterNotEqOperator": "Is not equal to",
  "grid.filterIsNullOperator": "Is null",
  "grid.filterIsNotNullOperator": "Is not null",
  "grid.filterIsEmptyOperator": "Is empty",
  "grid.filterIsNotEmptyOperator": "Is not empty",
  "grid.filterStartsWithOperator": "Starts with",
  "grid.filterContainsOperator": "Contains",
  "grid.filterNotContainsOperator": "Does not contain",
  "grid.filterEndsWithOperator": "Ends with",
  "grid.filterGteOperator": "Is greater than or equal to",
  "grid.filterGtOperator": "Is greater than",
  "grid.filterLteOperator": "Is less than or equal to",
  "grid.filterLtOperator": "Is less than",
  "grid.filterIsTrue": "Is true",
  "grid.filterIsFalse": "Is false",
  "grid.filterBooleanAll": "(All)",
  "grid.filterAfterOrEqualOperator": "Is after or equal to",
  "grid.filterAfterOperator": "Is after",
  "grid.filterBeforeOperator": "Is before",
  "grid.filterBeforeOrEqualOperator": "Is before or equal to",
  "grid.filterAndLogic": "And",
  "grid.filterOrLogic": "Or"
};
export {
  ie as aIResponseData,
  ee as adaptiveColumnMenuCheckboxFilterTitle,
  Z as adaptiveColumnMenuChooserSubTitle,
  Y as adaptiveColumnMenuChooserTitle,
  Q as adaptiveColumnMenuFilterTitle,
  re as adaptiveToolbarGroupTitle,
  te as adaptiveToolbarSortTitle,
  _ as columnChooserApplyButton,
  $ as columnChooserResetButton,
  C as columnMenu,
  X as columnMenuColumnChooserSelectedItems,
  V as columnMenuColumnChooserTitle,
  j as detailCollapse,
  U as detailExpand,
  M as editDialogCancelButtonTitle,
  G as editDialogSaveButtonTitle,
  R as editDialogTitle,
  D as exportPDF,
  f as filterApplyButton,
  w as filterAriaLabel,
  F as filterCheckAll,
  v as filterChooseOperator,
  h as filterClearAllButton,
  m as filterClearButton,
  A as filterResetButton,
  x as filterSelectAll,
  E as filterSelectedItems,
  b as filterSubmitButton,
  I as filterTitle,
  ae as generatedWithAI,
  J as gridAriaLabel,
  K as gridRowReorderAriaLabel,
  z as groupApplyButton,
  W as groupClearButton,
  k as groupCollapse,
  u as groupColumn,
  N as groupExpand,
  q as groupPanelAriaLabel,
  p as groupPanelEmpty,
  ue as messages,
  e as noRecords,
  r as pagerFirstPage,
  t as pagerInfo,
  a as pagerItemPerPage,
  i as pagerLastPage,
  l as pagerNextPage,
  g as pagerOf,
  n as pagerPage,
  s as pagerPageSizeAriaLabel,
  o as pagerPreviousPage,
  d as pagerTotalPages,
  O as searchPlaceholder,
  y as searchboxPlaceholder,
  H as selectRow,
  B as sortApplyButton,
  L as sortAriaLabel,
  P as sortAscending,
  T as sortClearButton,
  S as sortDescending,
  le as toolbarAI,
  ne as toolbarAIApply,
  pe as toolbarCheckboxFilter,
  de as toolbarColumnsChooser,
  ge as toolbarFilter,
  se as toolbarGroup,
  oe as toolbarSort,
  c as ungroupColumn
};
