// @progress/kendo-react-dropdowns/messages/index.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
const o = "dropdowns.nodata", r = "dropdowns.clear", a = "dropdowns.comboArrowBtnAriaLabelExpand", e = "dropdowns.comboArrowBtnAriaLabelCollapse", n = "dropdowns.dropDownListArrowBtnAriaLabel", d = "dropdowns.dropDownListAdaptiveFilterPlaceholder", t = "dropdowns.apply", l = "dropdowns.cancel", p = {
  [r]: "clear",
  [o]: "NO DATA FOUND.",
  [a]: "expand combobox",
  [e]: "collapse combobox",
  [n]: "select",
  [d]: "Filter",
  [t]: "Apply",
  [l]: "Cancel"
};
export {
  t as adaptiveModeFooterApply,
  l as adaptiveModeFooterCancel,
  r as clear,
  e as comboArrowBtnAriaLabelCollapse,
  a as comboArrowBtnAriaLabelExpand,
  d as dropDownListAdaptiveFilterPlaceholder,
  n as dropDownListArrowBtnAriaLabel,
  p as messages,
  o as nodata
};
