// @progress/kendo-react-grid/cells/editcell/utils.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { GRID_COL_INDEX_ATTRIBUTE as c } from "../../constants/index.mjs";
import { uGrid as a, classNames as i } from "@progress/kendo-react-common";
const o = (e) => {
  let t = null;
  const l = e.unstyled, d = l && l.uGrid ? l.uGrid : a, s = i(d.editTd({ selected: e.isSelected }), e.className);
  return t = {
    colSpan: e.colSpan,
    style: e.style,
    className: s,
    "aria-colindex": e.ariaColumnIndex,
    "aria-selected": e.isSelected,
    [c]: e.columnIndex,
    role: "gridcell"
  }, { tdProps: t };
}, u = (e) => {
  var t, l, d;
  return ((l = (t = e.cells) == null ? void 0 : t.edit) == null ? void 0 : l[e.editor || "text"]) || ((d = e.cells) == null ? void 0 : d.data);
};
export {
  u as getCustomCell,
  o as getEditCellTdProps
};
