// @progress/kendo-react-grid/columnMenu/GridColumnMenuFilterUI.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as e from "react";
import { GridColumnMenuFilterCell as l } from "./GridColumnMenuFilterCell.mjs";
import { GridContext as d } from "../utils/GridContext.mjs";
import { ButtonGroup as s, Button as m } from "@progress/kendo-react-buttons";
const E = (i) => {
  const { firstFilterProps: a, hideSecondFilter: c, secondFilterProps: u, logicValue: t, onLogicChange: r } = i, o = e.useContext(d), n = (t == null ? void 0 : t.operator) === "and";
  return /* @__PURE__ */ e.createElement(e.Fragment, null, /* @__PURE__ */ e.createElement(l, { ...a }), !c && /* @__PURE__ */ e.createElement(e.Fragment, null, /* @__PURE__ */ e.createElement(s, null, /* @__PURE__ */ e.createElement(
    m,
    {
      type: "button",
      value: "and",
      size: o.mobileMode ? "large" : "medium",
      togglable: !0,
      selected: n,
      onClick: r
    },
    "And"
  ), /* @__PURE__ */ e.createElement(
    m,
    {
      type: "button",
      value: "or",
      size: o.mobileMode ? "large" : "medium",
      togglable: !0,
      selected: !n,
      onClick: r
    },
    "Or"
  )), /* @__PURE__ */ e.createElement(l, { ...u })));
};
export {
  E as GridColumnMenuFilterUI
};
