// @progress/kendo-react-conversational-ui/messages/index.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
const e = "newMessage.sendButton", s = "newMessage.optionsButton", a = "newMessage.messageInput", t = "avatar.alt", o = "messageList.ariaLabel", n = "aIPrompt.noOutputs", l = "aIPrompt.cancel", c = "chat.senderDeletedMessage", i = "chat.receiverDeletedMessage", d = "chat.copyNotification", p = "chat.downloadAllFiles", g = "chat.closeReply", r = "chat.collapseMessage", h = "chat.expandMessage", M = "chat.removeFile", m = "chat.fileActions", u = "chat.attachFile", w = "chat.downloadAll", v = {
  [e]: "Send",
  [s]: "Options",
  [a]: "Message",
  [t]: "Avatar",
  [o]: "Message list",
  [l]: "Cancel",
  [n]: "No output from AI available",
  [c]: "You removed this message",
  [i]: "The message was removed by its sender",
  [d]: "Text copied to clipboard",
  [p]: "Download all files",
  [g]: "Close reply",
  [r]: "Collapse message",
  [h]: "Expand message",
  [M]: "Remove file",
  [m]: "File actions for",
  [u]: "Attach file",
  [w]: "Download all"
};
export {
  l as aIPromptCancel,
  n as aIPromptNoOutputs,
  o as ariaLabelMessageList,
  u as attachFile,
  t as avatarAlt,
  d as chatCopyNotification,
  i as chatReceiverDeletedMessage,
  c as chatSenderDeletedMessage,
  g as closeReply,
  r as collapseMessage,
  w as downloadAll,
  p as downloadAllFiles,
  h as expandMessage,
  m as fileActions,
  v as messages,
  a as newMessageMessageInput,
  s as newMessageOptionsButton,
  e as newMessageSendButton,
  M as removeFile
};
