// @progress/kendo-react-conversational-ui/index.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import { ChatMessage as e } from "./chat/components/ChatMessage.mjs";
import { Chat as m } from "./chat/components/Chat.mjs";
import { SuggestionGroup as a } from "./chat/components/SuggestionsGroup.mjs";
import { HeroCard as f } from "./chat/components/templates/HeroCard.mjs";
import { DEFAULT_FIELD_MAPPING as i, extractFieldMappingFromProps as P, isStandardMessageFormat as A, mapDataArrayToMessages as I, mapDataToMessage as n, mapMessageToCustomFormat as u } from "./chat/utils/fieldMapping.mjs";
import { AIPrompt as g } from "./ai-prompt/AIPrompt.mjs";
import { PromptHeader as D } from "./ai-prompt/AIPromptHeader.mjs";
import { AIPromptFooter as M } from "./ai-prompt/AIPromptFooter.mjs";
import { AIPromptContent as C } from "./ai-prompt/AIPromptContent.mjs";
import { AIPromptView as T } from "./ai-prompt/views/AIPromptView.mjs";
import { AIPromptOutputView as h } from "./ai-prompt/views/AIPromptOutputView.mjs";
import { AIPromptCommandsView as G } from "./ai-prompt/views/AIPromptCommandsView.mjs";
import { AIPromptViewRender as L } from "./ai-prompt/views/AIPromptViewRender.mjs";
import { InlineAIPrompt as _ } from "./inline-ai-prompt/InlineAIPrompt.mjs";
import { commandsViewDefaults as N, outputViewDefaults as O, promptViewDefaults as R } from "./ai-prompt/views/constants.mjs";
export {
  g as AIPrompt,
  G as AIPromptCommandsView,
  C as AIPromptContent,
  M as AIPromptFooter,
  h as AIPromptOutputView,
  T as AIPromptView,
  L as AIPromptViewRender,
  m as Chat,
  e as ChatMessage,
  i as DEFAULT_FIELD_MAPPING,
  f as HeroCard,
  _ as InlineAIPrompt,
  D as PromptHeader,
  a as SuggestionGroup,
  N as commandsViewDefaults,
  P as extractFieldMappingFromProps,
  A as isStandardMessageFormat,
  I as mapDataArrayToMessages,
  n as mapDataToMessage,
  u as mapMessageToCustomFormat,
  O as outputViewDefaults,
  R as promptViewDefaults
};
