// @progress/kendo-react-grid/header/GridHeaderSelectionCell.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import * as t from "react";
import { GridColumnMenuWrapper as l } from "../columnMenu/GridColumnMenuWrapper.mjs";
import { selectRow as n, messages as m } from "../messages/index.mjs";
import { useLocalization as d } from "@progress/kendo-react-intl";
import { useId as u } from "@progress/kendo-react-common";
import { GridContext as s } from "../utils/GridContext.mjs";
const g = (e) => {
  const o = t.useContext(s), c = u(), a = d(), r = e.id || c;
  return [
    /* @__PURE__ */ t.createElement("span", { key: 0, className: "k-checkbox-wrap" }, /* @__PURE__ */ t.createElement(
      "input",
      {
        id: r,
        type: "checkbox",
        tabIndex: -1,
        className: "k-checkbox k-checkbox-md k-rounded-md",
        "aria-label": a.toLanguageString(n, m[n]),
        "aria-checked": e.selectionValue,
        checked: e.selectionValue,
        onChange: (i) => o.onHeaderSelectionChange({
          field: e.field,
          syntheticEvent: i
        })
      }
    )),
    e.columnMenuWrapperProps.columnMenu && /* @__PURE__ */ t.createElement(l, { ...e.columnMenuWrapperProps })
  ];
};
export {
  g as GridHeaderSelectionCell
};
