// @progress/kendo-react-form/index.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import { Field as a } from "./Field.mjs";
import { Form as r } from "./Form.mjs";
import { FieldArray as i } from "./FieldArray.mjs";
import { FieldWrapper as l } from "./FieldWrapper.mjs";
import { FormElement as s } from "./FormElement.mjs";
import { withIdHOC as o } from "@progress/kendo-react-common";
import { FormSeparator as C } from "./FormSeparator.mjs";
import { FormFieldSet as y } from "./FormFieldSet.mjs";
const m = o(r);
m.displayName = "KendoReactForm";
export {
  a as Field,
  i as FieldArray,
  l as FieldWrapper,
  m as Form,
  r as FormClassComponent,
  s as FormElement,
  y as FormFieldSet,
  C as FormSeparator
};
