// @progress/kendo-react-grid/cells/groupcell/GridGroupCellServer.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as r from "react";
import { GridGroupCellServerContainer as c } from "./GridGroupCellServerContainer.mjs";
import { getClientCellProps as u, isClientReference as a } from "../../utils/index.mjs";
import { getGroupCellTdProps as d, getCustomCell as g } from "./utils.mjs";
const E = (i) => {
  const { cellProps: t } = i, l = u(t), { tdProps: e, td2Props: o, content: n, renderToggleCell: C } = d(t), s = g(t);
  if (s) {
    const p = a(s);
    return /* @__PURE__ */ r.createElement(
      c,
      {
        cellProps: l,
        tdProps: e,
        td2Props: o,
        isCustom: !0,
        isClient: p,
        renderToggleCell: !0
      },
      /* @__PURE__ */ r.createElement(s, { ...l, tdProps: e, td2Props: o }, n)
    );
  }
  const m = e ? /* @__PURE__ */ r.createElement("td", { ...e, key: e == null ? void 0 : e.key }, n) : null;
  return /* @__PURE__ */ r.createElement(
    c,
    {
      cellProps: l,
      renderToggleCell: C,
      tdProps: e,
      td2Props: o,
      content: n
    },
    m
  );
};
export {
  E as GridGroupCellServer
};
